### PR TITLE
library expansion and compatibility with other ZhianTec fp sensors

### PR DESCRIPTION
**Pull request by stitchesnburns**
this should address compatibility for other models of the same brand (ZhianTec) of fingerprint sensors, particularly models with higher memory capacity.  tested on a same-brand fingerprint sensor with 1000 template memory.

also, expanded the library by implementing methods such as read_sysparam(), get_fpdata(), send_fpdata(), empty_library(), as well as sending and receiving data payload.  fingerprint templates and images may now be downloaded/uploaded.  special note on the uploading/downloading of fingerprint images, data stream should not be taken as is and must be manipulated as stated in the manual (4.2.1).

### CODE REVIEW

File: examples/fingerprint_simpletest.py [hunk 1]
```diff
@@ -1,17 +1,19 @@
...
+#import busio
 from digitalio import DigitalInOut, Direction
 import adafruit_fingerprint
+import serial
 
 led = DigitalInOut(board.D13)
 led.direction = Direction.OUTPUT
 
-uart = busio.UART(board.TX, board.RX, baudrate=57600)
+#uart = busio.UART(board.TX, board.RX, baudrate=57600)
```
Comment (ladyada) [inline examples/fingerprint_simpletest.py]: don't change the default example setup

File: examples/fingerprint_simpletest.py [hunk 2]
```diff
@@ -1,17 +1,19 @@
 import time
 import board
-import busio
+#import busio
```
Comment (ladyada) [inline examples/fingerprint_simpletest.py]: don't change the default example setup

File: examples/fingerprint_simpletest.py [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
 from digitalio import DigitalInOut, Direction
 import adafruit_fingerprint
+import serial
```
Comment (ladyada) [inline examples/fingerprint_simpletest.py]: don't change the default example setup

File: examples/fingerprint_simpletest.py [hunk 4]
```diff
@@ -1,17 +1,19 @@
...
 led = DigitalInOut(board.D13)
 led.direction = Direction.OUTPUT
 
-uart = busio.UART(board.TX, board.RX, baudrate=57600)
+#uart = busio.UART(board.TX, board.RX, baudrate=57600)
 
 # If using with a computer such as Linux/RaspberryPi, Mac, Windows...
 #import serial
 #uart = serial.Serial("/dev/ttyUSB0", baudrate=57600, timeout=1)
+uart = serial.Serial("/dev/ttyAMA0", baudrate=57600, timeout=1)
```
Comment (ladyada) [inline examples/fingerprint_simpletest.py]: remove

File: examples/fingerprint_simpletest.py [hunk 5]
```diff
@@ -166,9 +171,14 @@ def get_num():
     if finger.read_templates() != adafruit_fingerprint.OK:
         raise RuntimeError('Failed to read templates')
     print("Fingerprint templates:", finger.templates)
+    if finger.count_templates() != adafruit_fingerprint.OK:
+        raise RuntimeError('Failed to read templates')
+    print("Number of templates: ", finger.template_count)
+    print("e) enroll print")
```
Comment (ladyada) [inline examples/fingerprint_simpletest.py]: remove

**Review by ladyada — CHANGES_REQUESTED** [review]
please make a new example as you like - dont change the default one

**Comment by stitchesnburns** [comment]
copy that.  will be creating a new example based on the default one.  will also rebase/update forked repo.

**Comment by stitchesnburns** [comment]
added saving the fingerprint image to a file in the example in response to @mhsattarian's [issue](https://github.com/adafruit/Adafruit_CircuitPython_Fingerprint/issues/15).  note that the fingerprint module is connected to RPi via GPIO.  connection via USB is not tested.

File: examples/fingerprint_simpletest.py (hunk 2 above)
Comment (stitchesnburns) [reply examples/fingerprint_simpletest.py]: reverted the default example and created a new one.

File: examples/fingerprint_simpletest.py (hunk 3 above)
Comment (stitchesnburns) [reply examples/fingerprint_simpletest.py]: reverted the default example and created a new one.

File: examples/fingerprint_simpletest.py (hunk 1 above)
Comment (stitchesnburns) [reply examples/fingerprint_simpletest.py]: reverted the default example and created a new one.

**Comment by mhsattarian** [comment]
Thanks a lot for this. I tested using `CH340G` USB to TTL converter and was able to get the image and save it using `OpenCV`. I'll test your code today, it should work perfectly.

**Comment by mhsattarian** [comment]
Hi, I just tested your code using the `CH340G` USB to TTL converter and it worked easily.

**Comment by FoamyGuy** [comment]
I don't have access to the other sensors that this is adding support for. But I have tested these changes with https://www.adafruit.com/product/751 Using the simpletest code: enrolling, finding, and deleting fingerprints is working as expected using the patched library.

**Comment by kattni** [comment]
@FoamyGuy Excellent, thank you! Please feel free to merge this PR.

**Comment by mhsattarian** [comment]
> I don't have access to the other sensors that this is adding support for. But I have tested these changes with [adafruit.com/product/751](https://www.adafruit.com/product/751) Using the simpletest code: enrolling, finding, and deleting fingerprints is working as expected using the patched library.

I had checked using the [R301T sensor](https://www.amazon.com/GROW-R301T-Biometrics-Fingerprint-Control/dp/B075167PMB). I know it's not from Adafruit, just wanted to let you know.

**Comment by FoamyGuy** [comment]
@mhsattarian Thank you for the link.

**Pull request merged**